include/boost/compute/algorithm/detail/find_extrema_reduce.hpp

```diff
@@ -0,0 +1,245 @@
+        "block_idx[lid] = acc_idx;\n" <<
+        "barrier(CLK_LOCAL_MEM_FENCE);\n" <<
+
+        k.decl<uint_>("group_offset") << " = count - (get_local_size(0) * get_group_id(0));\n";
+
+    k <<
+        "#pragma unroll\n"
+        "for(" << k.decl<uint_>("offset") << " = " << work_group_size << " / 2; offset > 0; " <<
+             "offset = offset / 2) {\n" <<
+             "if((lid < offset) && ((lid + offset) < group_offset)) { \n" <<
+                 k.decl<input_type>("mine") << " = block[lid];\n" <<
+                 k.decl<input_type>("other") << " = block[lid+offset];\n" <<
+                 "compare_result = mine " << sign << " other;\n" <<
+                 "block[lid] = compare_result ? mine : other;\n" <<
+                 "block_idx[lid] = compare_result ? " <<
+                     "block_idx[lid] : block_idx[lid+offset];\n" <<
+             "}\n"
+             "barrier(CLK_LOCAL_MEM_FENCE);\n" <<
+        "}\n" <<
+
+         // write block result to global output
+        "if(lid == 0){\n" <<
+        "    output[get_group_id(0)] = block[0];\n" <<
+        "    output_idx[get_group_id(0)] = block_idx[0];\n" <<
+        "}";
+
+    kernel kernel = k.compile(context);
+    kernel.set_arg(count_arg, static_cast<uint_>(count));
+    kernel.set_arg(output_arg, result.get_buffer());
+    kernel.set_arg(output_idx_arg, result_idx.get_buffer());
+    kernel.set_arg(block_arg, local_buffer<input_type>(work_group_size));
+    kernel.set_arg(block_idx_arg, local_buffer<uint_>(work_group_size));
+
+    queue.enqueue_1d_range_kernel(kernel,
+                                  0,
+                                  work_groups_no * work_group_size,
+                                  work_group_size);
+
+    return 0;
+}
+
+template<class InputIterator>
+InputIterator find_extrema_reduce(InputIterator first,
+                                  InputIterator last,
+                                  size_t preferred_work_group_size,
+                                  char sign,
+                                  command_queue &queue)
+{
+    typedef typename std::iterator_traits<InputIterator>::difference_type difference_type;
+    typedef typename std::iterator_traits<InputIterator>::value_type input_type;
+
+    const context &context = queue.get_context();
+    const device &device = queue.get_device();
+
+    // Getting information about used queue and device
+    const size_t compute_units_no = device.get_info<CL_DEVICE_MAX_COMPUTE_UNITS>();
+    const size_t max_work_group_size = device.get_info<CL_DEVICE_MAX_WORK_GROUP_SIZE>();
+    const bool is_queue_ordered =
+        static_cast<bool>(
+            queue.get_properties() &
+            command_queue::enable_out_of_order_execution
+        );
+
+    size_t count = detail::iterator_range_size(first, last);
+    size_t work_group_size = (std::min)(max_work_group_size, preferred_work_group_size);
+    size_t work_groups_no = compute_units_no * 64;
+    work_groups_no = (std::min)(
```

Comment:
For these sort of calculations, you may be able to use the `calculate_work_size()` function in [`<boost/compute/detail/work_size.hpp>`](https://github.com/boostorg/compute/blob/master/include/boost/compute/detail/work_size.hpp).